### PR TITLE
Update C# FFI

### PR DESCRIPTION
**Pull request by FreePhoenix888**

### CODE REVIEW

File: csharp/Platform.Data.Doublets.Benchmarks/MemoryBenchmarks.cs [hunk 1]
```diff
@@ -28,13 +30,17 @@ public static void Setup()
...
+            _ffiUnitedMemory = new FFI.UnitedMemoryLinks<uint>("db.links");
+            _ffiUnitedMemoryLinks = _ffiUnitedMemory.DecorateWithAutomaticUniquenessAndUsagesResolution();
         }
 
         [GlobalCleanup]
         public static void Cleanup()
         {
             _splitMemory.Dispose();
             _unitedMemory.Dispose();
+            _ffiUnitedMemory.Dispose();
```
Comment (uselessgoddess) [inline csharp/Platform.Data.Doublets.Benchmarks/MemoryBenchmarks.cs:43]: Wipe up after yourself. This isn't RAII.
```suggestion
            _ffiUnitedMemory.Dispose();
            File.Delete("db.links");
```

File: csharp/Platform.Data.Doublets.Tests/GenericLinksTests.cs [hunk 2]
```diff
@@ -41,6 +42,12 @@ private static void Using<TLink>(Action<ILinks<TLink>> action)
             {
                 action(scope.Use<ILinks<TLink>>());
             }
+            File.Delete("db.links");
+
+            using (var links = new FFI.UnitedMemoryLinks<TLink>("db.links"))
+            {
+                action(links);
+            }
```
Comment (uselessgoddess) [inline csharp/Platform.Data.Doublets.Tests/GenericLinksTests.cs:50]: May be
```suggestion
            using var links = new FFI.UnitedMemoryLinks<TLink>("db.links");
            action(links);
```

File: csharp/Platform.Data.Doublets.Tests/ILinksBasicTests.cs [hunk 3]
```diff
@@ -26,6 +26,26 @@ public static void DeleteAllUsages()
...
+            var a = links.CreatePoint();
+            var b = links.CreatePoint();
+
+            links.CreateAndUpdate(a, root);
+            links.CreateAndUpdate(b, root);
+
+            Assert.Equal(5U, links.Count());
+
+            links.DeleteAllUsages(root);
+
```
Comment (uselessgoddess) [inline csharp/Platform.Data.Doublets.Tests/ILinksBasicTests.cs:48]: It's a wrong test
Need `Assert.Equal(3U, links.Count())`

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    {
+        private static readonly UncheckedConverter<byte, TLink> from_u8 = UncheckedConverter<byte, TLink>.Default;
+        private static readonly UncheckedConverter<ushort, TLink> from_u16 = UncheckedConverter<ushort, TLink>.Default;
+        private static readonly UncheckedConverter<uint, TLink> from_u32 = UncheckedConverter<uint, TLink>.Default;
+        private static readonly UncheckedConverter<ulong, TLink> from_u64 = UncheckedConverter<ulong, TLink>.Default;
+        private static readonly UncheckedConverter<TLink, ulong> from_t = UncheckedConverter<TLink, ulong>.Default;
+
+        public LinksConstants<TLink> Constants { get; }
+
+        private readonly unsafe void* _body;
```
Comment (uselessgoddess) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:147]: May be
```suggestion
        private readonly unsafe void* _ptr;
```

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+                    {
+                        throw new NotImplementedException();
+                    }
+                }
+            }
+        }
+
+        protected override void Dispose(bool manual, bool wasDisposed)
+        {
+            if (wasDisposed)
```
Comment (uselessgoddess) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:394]: Don't shoot yourself in the foot
```suggestion
            if (wasDisposed && _body != null)
```

**Review by uselessgoddess — CHANGES_REQUESTED** [review]
Look here

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+using System.Linq;
+using System.Runtime.InteropServices;
+using Platform.Converters;
+using Platform.Disposables;
+using IDisposable = Platform.Disposables.IDisposable;
+using static System.Runtime.CompilerServices.Unsafe;
+
+namespace Platform.Data.Doublets.FFI
+{
+    struct FfiLink_Uint8
```
Comment (Konard) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:12]:
```suggestion
    struct FfiLink_UInt8
```

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+    {
+        public Byte Index;
+        public Byte Source;
+        public Byte Target;
+    }
+
+    struct FfiLink_Uint16
```
Comment (Konard) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:19]:
```suggestion
    struct FfiLink_UInt16
```

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+        public static extern ushort UInt16UnitedMemoryLinks_Each(void* self, EachCallback_Uint16 callback, ushort* restriction, nuint len);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern uint UInt32UnitedMemoryLinks_Each(void* self, EachCallback_Uint32 callback, uint* restriction, nuint len);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern ulong UInt64UnitedMemoryLinks_Each(void* self, EachCallback_Uint64 callback, ulong* restriction, nuint len);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern byte ByteUnitedMemoryLinks_Update(void* self, byte* restriction, nuint restrictionsLength,  byte* substitution, nuint substitutionLength);
```
Comment (Konard) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:113]:
```suggestion
        public static extern byte ByteUnitedMemoryLinks_Update(void* self, byte* restriction, nuint restrictionLength, byte* substitution, nuint substitutionLength);
```

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        public static extern ushort UInt16UnitedMemoryLinks_Update(void* self, ushort* restriction, nuint restrictionsLength,  ushort* substitution, nuint substitutionLength);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern uint UInt32UnitedMemoryLinks_Update(void* self, uint* restriction, nuint restrictionsLength,  uint* substitution, nuint substitutionLength);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern ulong UInt64UnitedMemoryLinks_Update(void* self, ulong* restriction, nuint restrictionsLength,  ulong* substitution, nuint substitutionLength);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern byte ByteUnitedMemoryLinks_Delete(void* self, byte* restriction, nuint len);
```
Comment (Konard) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:125]:
```suggestion
        public static extern byte ByteUnitedMemoryLinks_Delete(void* self, byte* restriction, nuint restrictionLength);
```

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+        public static extern void UInt16UnitedMemoryLinks_Drop(void* self);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern void UInt32UnitedMemoryLinks_Drop(void* self);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern void UInt64UnitedMemoryLinks_Drop(void* self);
+
+        [DllImport(DllName, CallingConvention = CallingConvention.Cdecl)]
+        public static extern byte ByteUnitedMemoryLinks_Create(void* self);
```
Comment (Konard) [inline csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:77]:
```suggestion
        public static extern byte ByteUnitedMemoryLinks_Create(void* self, byte* substitution, nuint substitutionLength);
```

**Review by Konard — CHANGES_REQUESTED** [review]

File: csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs (hunk 9 above)
Comment (uselessgoddess) [reply csharp/Platform.Data.Doublets/FFI/UnitedMemoryLinks.cs:125]: 🥺

**Review by uselessgoddess — APPROVED** [review]

**Pull request merged**